File: sample/src/main/java/com/afollestad/iconrequestsample/MainActivity.java
package com.afollestad.iconrequestsample;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;

import android.annotation.SuppressLint;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v4.content.FileProvider;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;
import com.afollestad.assent.Assent;
import com.afollestad.assent.AssentActivity;
import com.afollestad.iconrequest.PolarConfig;
import com.afollestad.iconrequest.PolarRequest;
import com.afollestad.materialdialogs.MaterialDialog;
import io.reactivex.disposables.CompositeDisposable;
import java.io.File;
import me.zhanghai.android.materialprogressbar.MaterialProgressBar;

public class MainActivity extends AssentActivity implements Toolbar.OnMenuItemClickListener {

  @BindView(R.id.rootView)
  View rootView;

  @BindView(R.id.progress)
  MaterialProgressBar progressView;

  @BindView(R.id.fab)
  FloatingActionButton fabView;

  @BindView(R.id.toolbar)
  Toolbar toolbar;

  @BindView(R.id.list)
  RecyclerView listView;

  private PolarRequest request;
  private Unbinder unbinder;
  private MainAdapter adapter;
  private MaterialDialog dialog;
  private CompositeDisposable subs;

  @OnClick(R.id.fab)
  public void onClickFab() {
    if (!Assent.isPermissionGranted(Assent.WRITE_EXTERNAL_STORAGE)) {
      Assent.requestPermissions(
          results -> {
            if (results.allPermissionsGranted()) {
              request.send().subscribe();
            } else {
              Snackbar.make(rootView, R.string.permission_denied, Snackbar.LENGTH_LONG).show();
            }
          },
          69,
          Assent.WRITE_EXTERNAL_STORAGE);
      return;
    }
    request.send().subscribe();
  }

  @SuppressLint("InlinedApi")
  @Override
  protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);
    unbinder = ButterKnife.bind(this);

    subs = new CompositeDisposable();

    toolbar.inflateMenu(R.menu.menu_main);
    toolbar.setOnMenuItemClickListener(this);
    fabView.hide();

    adapter = new MainAdapter();
    adapter.setListener((index, app) -> request.toggleSelection(app));

    GridLayoutManager lm =
        new GridLayoutManager(this, getResources().getInteger(R.integer.grid_width));
    listView.setLayoutManager(lm);
    listView.setAdapter(adapter);

    PolarConfig config =
        PolarConfig.create(this).emailRecipient("fake-email@helloworld.com").build();
    request =
        PolarRequest.make(this, savedInstanceState)
            .config(config)
            .uriTransformer(
                uri ->
                    FileProvider.getUriForFile(
                        MainActivity.this,
                        BuildConfig.APPLICATION_ID + ".fileProvider",
                        new File(uri.getPath())));

    subs.add(
        request
            .loading()
            .subscribe(isLoading -> progressView.setVisibility(isLoading ? VISIBLE : GONE)));
    subs.add(
        request
            .loaded()
            .subscribe(
                loadResult -> {
                  if (!loadResult.success()) {
                    adapter.setAppsList(null);
                    loadResult.error().printStackTrace();
                    Snackbar.make(rootView, loadResult.error().getMessage(), Snackbar.LENGTH_LONG)
                        .show();
                    return;
                  }
                  adapter.setAppsList(loadResult.apps());
                  invalidateToolbar();
                }));
    subs.add(
        request
            .selectionChange()
            .subscribe(
                appModel -> {
                  adapter.update(appModel);
                  invalidateToolbar();
                }));
    subs.add(
        request
            .sending()
            .subscribe(
                isSending -> {
                  if (isSending) {
                    dialog =
                        new MaterialDialog.Builder(MainActivity.this)
                            .content(R.string.preparing_your_request)
                            .progress(true, -1)
                            .cancelable(false)
                            .canceledOnTouchOutside(false)
                            .show();
                  } else if (dialog != null) {
                    dialog.dismiss();
                  }
                }));
    subs.add(
        request
            .sent()
            .subscribe(
                sendResult -> {
                  if (!sendResult.success()) {
                    sendResult.error().printStackTrace();
                    Snackbar.make(rootView, sendResult.error().getMessage(), Snackbar.LENGTH_LONG)
                        .show();
                  } else {
                    Snackbar.make(rootView, R.string.request_sent, Snackbar.LENGTH_SHORT).show();
                  }
                }));

    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      int flags = getWindow().getDecorView().getSystemUiVisibility();
      flags |= View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR;
      getWindow().getDecorView().setSystemUiVisibility(flags);
    }
  }

  @Override
  protected void onResume() {
    super.onResume();
    if (request.getLoadedApps().isEmpty()) {
      request.load().subscribe();
    }
  }

  private void invalidateToolbar() {
    subs.add(
        request
            .getSelectedApps()
            .subscribe(
                appModels -> {
                  int selectedCount = appModels.size();
                  if (selectedCount == 0) {
                    fabView.hide();
                    toolbar.setTitle(R.string.app_name);
                    toolbar
                        .getMenu()
                        .findItem(R.id.selectAllNone)
                        .setIcon(R.drawable.ic_action_selectall);
                  } else {
                    fabView.show();
                    toolbar.setTitle(getString(R.string.app_name_x, selectedCount));
                    toolbar
                        .getMenu()
                        .findItem(R.id.selectAllNone)
                        .setIcon(R.drawable.ic_action_selectall);
                  }
                }));
  }

  @Override
  protected void onSaveInstanceState(Bundle outState) {
    request.saveInstance(outState);
    super.onSaveInstanceState(outState);
  }

  @Override
  protected void onDestroy() {
    super.onDestroy();
    subs.dispose();
    unbinder.unbind();
    request = null;
  }

  @Override
  protected void onPause() {
    super.onPause();
    if (dialog != null) {
      dialog.dismiss();
      dialog = null;
    }
  }

  @Override
  public boolean onMenuItemClick(MenuItem item) {
    if (item.getItemId() == R.id.selectAllNone) {
      if (!request.getSelectedApps().blockingGet().isEmpty()) {
        request.deselectAll();
        item.setIcon(R.drawable.ic_action_selectall);
      } else {
        request.selectAll();
        item.setIcon(R.drawable.ic_action_selectnone);
      }
      return true;
    }
    return false;
  }
}

File: sample/src/main/java/com/afollestad/iconrequestsample/MainActivity.kt
package com.afollestad.iconrequestsample

import android.Manifest.permission.WRITE_EXTERNAL_STORAGE
import android.content.pm.PackageManager
import android.os.Build.VERSION.SDK_INT
import android.os.Build.VERSION_CODES
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.support.v4.content.FileProvider
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.Toolbar
import android.view.MenuItem
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import com.afollestad.iconrequest.ArcticConfig
import com.afollestad.iconrequest.ArcticRequest
import com.afollestad.iconrequest.UriTransformer
import kotlinx.android.synthetic.main.activity_main.fab
import kotlinx.android.synthetic.main.activity_main.list
import kotlinx.android.synthetic.main.activity_main.progress
import kotlinx.android.synthetic.main.activity_main.rootView
import kotlinx.android.synthetic.main.activity_main.toolbar
import java.io.File

class MainActivity : AppCompatActivity(), Toolbar.OnMenuItemClickListener {

  companion object {
    private const val WRITE_EXTERNAL_STORAGE_RQ = 69
  }

  private lateinit var adapter: MainAdapter
  private lateinit var request: ArcticRequest

  private fun onClickFab() {
    val permissionGrantedOrNot = ContextCompat.checkSelfPermission(this, WRITE_EXTERNAL_STORAGE)
    if (permissionGrantedOrNot == PackageManager.PERMISSION_GRANTED) {
      request.performSend()
    } else {
      ActivityCompat.requestPermissions(
          this, arrayOf(WRITE_EXTERNAL_STORAGE), WRITE_EXTERNAL_STORAGE_RQ
      )
    }
  }

  override fun onRequestPermissionsResult(
    requestCode: Int,
    permissions: Array<out String>,
    grantResults: IntArray
  ) {
    super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    if (requestCode == WRITE_EXTERNAL_STORAGE_RQ && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
      onClickFab()
    } else {
      Snackbar.make(rootView, R.string.permission_denied, Snackbar.LENGTH_LONG)
          .show()
    }
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)

    toolbar.inflateMenu(R.menu.menu_main)
    toolbar.setOnMenuItemClickListener(this)

    fab.hide()
    fab.setOnClickListener { onClickFab() }

    adapter = MainAdapter()
    adapter.setListener { _, app -> request.toggleSelection(app) }

    val lm = GridLayoutManager(this, resources.getInteger(R.integer.grid_width))
    list.layoutManager = lm
    list.adapter = adapter

    val config =
      ArcticConfig(
          emailRecipient = "fake-email@helloworld.com",
          errorOnInvalidDrawables = false,
          apiHost = "https://arcticmanager.com",
          apiKey = "fb964410-0230-4f2b-ac32-4625f3c2929c"
      )
    val uriTransformer: UriTransformer = {
      FileProvider.getUriForFile(
          this@MainActivity,
          BuildConfig.APPLICATION_ID + ".fileProvider",
          File(it.path)
      )
    }

    request = ArcticRequest(
        context = this,
        savedInstanceState = savedInstanceState,
        config = config,
        uriTransformer = uriTransformer,
        onLoading = { progress.visibility = VISIBLE },
        onLoadError = {
          progress.visibility = GONE
          adapter.setAppsList(null)
          Snackbar.make(rootView, it.message!!, Snackbar.LENGTH_LONG)
              .show()
        },
        onLoaded = {
          progress.visibility = GONE
          adapter.setAppsList(it)
          invalidateToolbar()
        },
        onSelectionChange = {
          adapter.update(it)
          invalidateToolbar()
        },
        onSending = {
          progress.visibility = VISIBLE
        },
        onSendError = {
          progress.visibility = GONE
          Snackbar.make(rootView, it.message!!, Snackbar.LENGTH_LONG)
              .show()
        },
        onSent = {
          progress.visibility = GONE
          Snackbar.make(rootView, R.string.request_sent, Snackbar.LENGTH_SHORT)
              .show()
        }
    )

    if (SDK_INT >= VERSION_CODES.O) {
      var flags = window.decorView.systemUiVisibility
      flags = flags or View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR
      window.decorView.systemUiVisibility = flags
    }
  }

  override fun onResume() {
    super.onResume()
    if (request.loadedApps.isEmpty()) {
      request.performLoad()
    }
  }

  override fun onPause() {
    request.dispose()
    super.onPause()
  }

  private fun invalidateToolbar() {
    val selectedCount = request.selectedApps.size
    if (selectedCount == 0) {
      fab.hide()
      toolbar.setTitle(R.string.app_name)
      toolbar.menu
          .findItem(R.id.selectAllNone)
          .setIcon(R.drawable.ic_action_selectall)
    } else {
      fab.show()
      toolbar.title = getString(R.string.app_name_x, selectedCount)
      toolbar.menu
          .findItem(R.id.selectAllNone)
          .setIcon(R.drawable.ic_action_selectall)
    }
  }

  override fun onSaveInstanceState(outState: Bundle) {
    request.saveInstance(outState)
    super.onSaveInstanceState(outState)
  }

  override fun onMenuItemClick(item: MenuItem): Boolean {
    if (item.itemId == R.id.selectAllNone) {
      if (!request.selectedApps.isEmpty()) {
        request.deselectAll()
        item.setIcon(R.drawable.ic_action_selectall)
      } else {
        request.selectAll()
        item.setIcon(R.drawable.ic_action_selectnone)
      }
      return true
    }
    return false
  }
}
